### PR TITLE
Iteration 8 - Exceptions

### PR DESCRIPTION
**Pull request by semitsvetik**
Iteration 2 - Test#4 is complete

### CODE REVIEW

**Comment by semitsvetik** [comment]
How can I remove all reviewed commits from Pull Request?
All commits before "Iteration 2 - Test#4 is complete" can be removed from PR.

**Comment by semitsvetik** [comment]
I am stuck in further design for iteration 5.
We need to delegate behaviour to Command class but how to do it without code duplication?

**Comment by semitsvetik** [comment]
Iteration 5 is complete. Code is ready.

File: src/main/java/com/acme/dbo/txlog/ByteCommand.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
+package com.acme.dbo.txlog;
+
+import static java.lang.Math.abs;
+
+public class ByteCommand {
+        private static String DECOR = "primitive: ";
+
+        public Byte currentValue;
```
Comment (eugene-krivosheyev) [inline src/main/java/com/acme/dbo/txlog/ByteCommand.java]: public!!!! ))) Please hide the state. No one likes see the internals )

File: src/main/java/com/acme/dbo/txlog/ByteCommand.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        private static String accumulator;
```
Comment (eugene-krivosheyev) [inline src/main/java/com/acme/dbo/txlog/ByteCommand.java]: static? Nope! we have to have _instance_ variables not class-wide variables. Your code works fine until we have to create more that one Controller instance. After it both of them actually will have one state.

File: src/main/java/com/acme/dbo/txlog/ByteCommand.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+                sum = this.currentValue;
+            } else {
+                if (checkByteValueIsOutBound((byte) (this.currentValue))) {
+                    Controller.flush();
+                    accumulator = MAX_BYTE + "";
+                    sum = MAX_BYTE;
+                    Controller.flush();
+                } else {
+                    sum = (byte) (sum + this.currentValue);
+                    if (accumulator.contains(System.lineSeparator())) {
```
Comment (eugene-krivosheyev) [inline src/main/java/com/acme/dbo/txlog/ByteCommand.java]: 3-level if and we need to go deeper ))) Please no ) Feel free to Extract methods, it won't hurt )

File: src/main/java/com/acme/dbo/txlog/Facade.java [hunk 4]
```diff
@@ -1,54 +1,52 @@
...
+        Controller.intCommand = new IntCommand(message);
+        Controller.log(Controller.intCommand);
     }
 
-    public static void log(boolean message) { Controller.logBoolean(message,  isDecorated); }
-
     public static void log(byte message) {
-        Controller.logByte(message,  isDecorated);
+        Controller.byteCommand = new ByteCommand(message);
+        Controller.log(Controller.byteCommand);
```
Comment (eugene-krivosheyev) [inline src/main/java/com/acme/dbo/txlog/Facade.java]: maybe inline variable? There is no value in separate variable so you can inline it:
`Controller.log(new ByteCommand());`

File: src/main/java/com/acme/dbo/txlog/ByteCommand.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+        public ByteCommand(Byte message) {
+            currentValue = message;
+            if (accumulator == null) {
+                accumulator = message.toString();
+                sum = message;
+            }
+        }
+
+        public static String getDecoratedState() {
```
Comment (eugene-krivosheyev) [inline src/main/java/com/acme/dbo/txlog/ByteCommand.java]: Static methods ((( With OOP we don't use it. For static methods won't work inheritance and overriding so you can't extend your system through inheritance. You can't subclass your ByteCommand in future ((( And as side effect it has very bad testability. So if we move to OOP you have to get rid of static methods and use _object_ state and methods.

File: src/main/java/com/acme/dbo/txlog/Facade.java [hunk 6]
```diff
@@ -1,54 +1,52 @@
...
-import java.util.Arrays;
-
-import static java.lang.Math.abs;
-
 public class Facade {
-    private static Boolean isDecorated=true;
 
     public static void log(int message) {
-    Controller.logInteger(message,  isDecorated);
+        Controller.intCommand = new IntCommand(message);
```
Comment (eugene-krivosheyev) [inline src/main/java/com/acme/dbo/txlog/Facade.java]: It's time to make Controller great again - make it true object ))) By replacing _static_ methods.

File: src/main/java/com/acme/dbo/txlog/PairValues.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package com.acme.dbo.txlog;
+
+public class PairValues {
```
Comment (eugene-krivosheyev) [inline src/main/java/com/acme/dbo/txlog/PairValues.java]: Oh, that's cool true object approach

File: src/main/java/com/acme/dbo/txlog/WriterFactory.java [hunk 8]
```diff
@@ -0,0 +1,8 @@
+package com.acme.dbo.txlog;
+
+public class WriterFactory {
```
Comment (eugene-krivosheyev) [inline src/main/java/com/acme/dbo/txlog/WriterFactory.java]: It's not Factory )) Factory _returns_ writer. Instead your factory do _write_ so have to be Writer

**Review by eugene-krivosheyev — CHANGES_REQUESTED** [review]

**Review by eugene-krivosheyev — CHANGES_REQUESTED** [review]
ask if any question

File: src/main/java/com/acme/dbo/txlog/Facade.java (hunk 4 above)
Comment (semitsvetik) [reply src/main/java/com/acme/dbo/txlog/Facade.java]: Would there be a link in this case inside Controller to ByteCommand?

File: src/main/java/com/acme/dbo/txlog/WriterFactory.java (hunk 8 above)
Comment (semitsvetik) [reply src/main/java/com/acme/dbo/txlog/WriterFactory.java]: Isn't it a Factory pattern? Why not naming it as Factory as we seemed to be doing so in our cases? We used to add pattern to our name. I also don't store link to Writer in Controller. Should i?

File: src/main/java/com/acme/dbo/txlog/Facade.java (hunk 4 above)
Comment (eugene-krivosheyev) [reply src/main/java/com/acme/dbo/txlog/Facade.java]: It depends on Controller's logic. If it decides to store it in variable or not (based on business logic of accumulating/flushing). From the outside we only have to give command and not to manage Controller's internal state. It have to be managed by Controller only.

File: src/main/java/com/acme/dbo/txlog/WriterFactory.java (hunk 8 above)
Comment (eugene-krivosheyev) [reply src/main/java/com/acme/dbo/txlog/WriterFactory.java]: Nope ) We name something as Factory only if _gives_ you some other objects. Your class doesn't give any objects, it just writes ) So it's Writer not WriterFactory.

File: src/main/java/com/acme/dbo/txlog/ByteCommand.java (hunk 5 above)
Comment (semitsvetik) [reply src/main/java/com/acme/dbo/txlog/ByteCommand.java]: I can't get rid of static methods. Removed all static methods and fields except static imports. Compiler keeps saying in Test files in rows:
 "        Facade.log(1);
        Facade.flush();"
Non-static method log (int) can't be referenced from static context. Where else is this static context setup? Idea suggests to make Facade.log and Facade.flush static for some reason.

Comment (eugene-krivosheyev) [reply src/main/java/com/acme/dbo/txlog/ByteCommand.java]: Sure you can ) We have made design decisions:
1. Facade stay with static methods for backward compatibility.
2. New classes (e.g. Controller) we design by OOP way: it won't have any static methods.
So your solution will look like:

```
class Facade {
  private static Control controller = new Controller(...anything it needs...); // <-- it's object!

  static void log(int message) {
    controller.log(new IntCommand(message)); // <-- you call _object_ method
  }
}
```

Comment (semitsvetik) [reply src/main/java/com/acme/dbo/txlog/ByteCommand.java]: Thank you! It worked!

File: src/main/java/com/acme/dbo/txlog/WriterFactory.java (hunk 8 above)
Comment (semitsvetik) [reply src/main/java/com/acme/dbo/txlog/WriterFactory.java]: Resolved.

File: src/main/java/com/acme/dbo/txlog/Facade.java (hunk 4 above)
Comment (semitsvetik) [reply src/main/java/com/acme/dbo/txlog/Facade.java]: Done

File: src/main/java/com/acme/dbo/txlog/ByteCommand.java (hunk 2 above)
Comment (semitsvetik) [reply src/main/java/com/acme/dbo/txlog/ByteCommand.java]: Done.

File: src/main/java/com/acme/dbo/txlog/ByteCommand.java (hunk 3 above)
Comment (semitsvetik) [reply src/main/java/com/acme/dbo/txlog/ByteCommand.java]: Ready

**Comment by semitsvetik** [comment]
There is a new PR with all corrections.

**Comment by semitsvetik** [comment]
Pull request is ready for review!

**Comment by semitsvetik** [comment]
Pull request is ready for review. Iteration 6 is complete with amendments.

**Comment by semitsvetik** [comment]
Request is ready for review. Final version. Iteration 6 with amendments.

**Pull request closed without merging**